Initial Draft for Issue 269 (#313)
canonical/open-documentation-academy: website/events/community-hour-discussions/odh-015-docs-as-code.md

> # Overview of Documentation Community Hour[ODH-015]: Documentation as Code

Reviewer: Suggestion and example:
For consistency, the article should follow how we usually format the file name, title, or the beginning of the document. I'll let you take a look at the rest of the document to see if anything else needs to be changed for consistency.

Author: Ok, will update this.

## Introduction

> Many of us in tech have experienced the challenge of writing both code and documentation, often favoring code over documentation. The philosophy of Documentation as Code, more commonly known as _**Docs as Code**_, bridges this gap by treating documentation with the similar tools, processes, and respect as source code. In this session, Robert Krátký explains how this approach streamlines documentation development by integrating it closely with source code in software projects.

Reviewer: Suggestion:
When you say it "bridges this gap", do you mean that it bridges the gap between writing code and documentation?
I would remove the "often favoring code over documentation", to make the introduction more concise and also because it might not be the case for some readers. What do you think?

```suggestion
The philosophy of Documentation as Code, more commonly known as _**Docs as Code**_, bridges the gap between software development and documentation by treating documentation with software tools and processes. In this session, Robert Krátký explains how this approach streamlines documentation development by integrating it closely with source code in software projects.
```

Watch the full video here: [UbuntuOnAir: ODH-015](https://www.youtube.com/watch?v=AVNfH99KiME)

---

## What is *Docs as Code*?

> *Docs as Code* is an approach that treats documentation as a first-class citizen alongside software source code. Instead of favoring code, it places documentation on equal footing in the software development lifecycle.

Reviewer: Question:
- What do you mean by "treats documentation as a first-class citizen"? Why have you decided to use that as the definition?
- Also, I don't think Docs as Code is about placing "documentation on equal footing in the software development lifecycle". I can happen, but I don't think it's one of the main reasons we use this approach. What do you think?

Author: Hi @Sophie-Pages , so in the video Robert mentions around the 5 minute mark that Docs as Code means treating Documentation same as software code. I just picked up on that point and try to emphasis with a simile. But, I can simplify this definition to make it more accessible. That is cool!

> This approach avoids reliance on a single monolithic system, it leverages specialized tools throughout the content creation, review, and publishing process. Often, the choice of tools aligns with the specific software process it addresses.

Reviewer: Suggestion:

I would suggest to either define what a "monolithic system" is or to redo the sentence without it to make it more accessible. Some users might not understand what a "single monolithic system" is. What do you think?

Author: Again, the "monolithic system" reference comes from the video. But, I will add a definition for accessibility.

Reviewer: Yes, you're absolutely right: it was shared without additional information in the video. In that case and to ensure our users can understand what we are explaining, I think it's fine to extend a bit Robert's video and ideas.

---

## Workflow

> One of the core objectives of the *Docs as Code* approach is to promote using the most suitable tools for each documentation task. As the saying goes, 'If your only tool is a hammer, every problem looks like a nail.' Understanding the purpose and selecting the right tools leads to more effective and valuable documentation.

Reviewer: Nitpick:

I would explain in more details the meaning and conclusion of the quote "If your only tool is a hammer, every problem looks like a nail" if you want to use it. 
Users might have enough information with the quote and the following sentence, but as they can come from any culture or background, I want to make sure the point is understandable for everyone.

Author: Sure. I can try and improve on this.

We can further understand this through the typical lifecycle of documentation when it is maintained as code.

| Goals                | Tools                                      | Reasoning                                          		      |
|----------------------|-------------------------------------------|------------------------------------------------------------------|
| Content Authoring    | Text Editors                              | Use non-binary, human-readable formats 			|
| Formatting & Styling | Markup Languages                          | Semantic formatting with stylesheets enables multiple output formats (HTML, PDF) |
| Version Control      | Git, SVN, Mercurial                       | Essential for tracking changes and history       					|
| Issue Tracking       | JIRA, Bugzilla, Mantis                    | Manage bugs and improvements                      |
| Testing & Validation | Scripts, Linters, Spell Checkers          | Borrow software testing concepts to ensure quality and reusability |

> | Publishing           | Static Site Generators, CMS               | Automated hosting and updates                      |

Reviewer: Suggestions
- Suggestion to clean the table (hopefully it should still work!).
- I haven't made the modifications, but what do you think of reviewing the case of some words? For example, I don't think "Issue Tracking", "Spell Checkers", or "Text Editors" need to be in title case. 
- Also, I don't think all our users know what CMS means. When I first use an acronym, I recommend that we spell it out and use the acronym in parentheses. The acronym can then be used throughout the document. Could you please review the document to define acronyms? Thank you!

```suggestion
| Goals | Tools | Reasoning |
|---|---|---|
| Content Authoring | Text Editors | Use non-binary, human-readable formats |
| Formatting & Styling | Markup Languages | Semantic formatting with stylesheets enables multiple output formats (HTML, PDF) |
| Version Control | Git, SVN, Mercurial | Essential for tracking changes and history |
| Issue Tracking | JIRA, Bugzilla, Mantis | Manage bugs and improvements |
| Testing & Validation | Scripts, Linters, Spell Checkers | Borrow software testing concepts to ensure quality and reusability |
| Publishing | Static Site Generators, CMS | Automated hosting and updates |
```

- Documentation tools support automated processes like change tracking, testing, and publishing via static site generators or content management systems ensuring high quality and seamless updates.
  
- Version control of documentation provides:
  - Parallel updates and easier collaboration.
  - Detailed change tracking that clearly matches modifications to their authors.
  - Improved transparency, flexibility, and maintainability across the project.
    
- It streamlines the integration and maintenance of documentation alongside code, for example:
	- Completion of functional documentation hand in hand with software development, synchronized with key project milestones such as software release.
	- Peer document reviews tied closely with code reviews.
	- Accurate document-to-code version alignment, keeping docs synced with the corresponding code on each individual software release branch.
   
- The core tools used are typically open-source and portable, rather than proprietary solutions (e.g., Adobe FrameMaker, MadCap, or RoboHelp). This avoids vendor lock-in, preventing dependency on a single vendor’s ecosystem that complicates switching or migration.
  

> - Developers already know these tools- including IDEs, text editors, and scripting tools- making it more natural for them to contribute to documentation, fostering stronger collaboration.

Reviewer: Suggestion and question:
I suggest separating this block of bullet points into different paragraphs or topics. If you find that it's possible, subheaders can be useful to separate it into different topics. 

You've used quite a few lists and bullet points in the last part of this article (here and below as well). I would like you to ask yourself: 
- Are all the lists needed? 
- Is there a better way to show the information?

Author: Hi @Sophie-Pages , There was a lot of disjoint information given in the video. I thought putting it in list would help readers with retention of the main points. But I can change the sub list to paragraphs. What do you think works better?

Reviewer: That makes a lot of sense. Lists are a great way to skim through information easily. In that case, I find that the nested list about the **Positives** is a bit too long and actually difficult to skim through. What I would recommend is separating this nested list into different topics like automation, version control, maintenance, and tools (and add those as headers). 
The **Negatives** and **Practical tips** we can keep them as they are. 
What do you think?

Author: Okay. Will do that.

> ### The Bad and the Ugly

Reviewer: Question and suggestion:
- Why have you decided to use "The good" and "The bad and the ugly" as headers?
- I recommend to put headers in sentence case (only the first letter is capitalized) if they don't relate to acronyms (Content Management Systems), product names (Ubuntu Pro), program names, etc. You can find more about this in the Canonical documentation style guide (https://docs.ubuntu.com/styleguide/en/#headings).

Author: Was just trying to make it an interesting read, by using a famous American movie reference. 
Will remove it and make it positives and negatives.

Reviewer: I like the idea of adding your personality inside the document! For that specific case, some users may not have the reference and may be taken aback by "The bad and the ugly". I think positives/negatives or benefits/challenges may be more understandable.

- Writers may face a steep learning curve adapting to markup languages and distributed tooling.
- Developers might resist added documentation responsibilities.
- Switching markup dialects (e.g., Markdown to reStructuredText) can be cumbersome.
- Finding user-friendly tools for build and integration is often difficult.
- Transitioning workflows may cause temporary release disruptions.

### Practical Tips

- Use portable and familiar editors for collaboration.
- Keep documentation in a version control system for easy tracking.
- Automated tests and checks can help catch discrepancies early, to the point of preventing release of incorrect software by identifying outdated docs.

---

## What else ...

### Upcoming Innovations

- "Diagrams as Code" is emerging to keep images and graphics updated alongside text and code.

### Other Callouts

1. Efforts are underway to streamline collaboration with external projects.
2. Automation is being developed to verify Contributor License Agreement(CLA) before contributors can submit pull requests to the Open Documentation Repository.
3. The Open Documentation schedule is available on the Forum to plan future Office Hour discussion topics.

> 4. Next week’s session will feature Nick Veitch discussing *Vale* for testing documentation.

Reviewer: Question:

This part doesn't relate to Documentation as Code. Why have you decided to include it?

Author: Hi @Sophie-Pages, I can remove this part if needed. I thought it was part of the initial Issue that any other callouts must be included too. Please comment if you want me to remove it.

Reviewer: I think it would be better to keep our article focused on the initial topic and to remove this. My initial issue might not have been clear enough, sorry about that. I think it's great to add other callouts, but only if they relate to the topic.

Author: Sure. Will do that.
